### ISSUE TITLE
RFC: support symlinks in `sgx.trusted_files`

### ISSUE THREAD

**Issue by woju** [gramineproject/gramine#1176]
### Description of the feature

# how it would work

1) `gramine-sgx-sign`, upon encountering a symlink, would emit a symlink specification (see below for options) instead of measuring the contents of the file.

2) We'd have some `realpath(3)` equivalent in trusted PAL that would resolve the file paths to a canonical form (which does not need to be the `realpath` outside gramine, but `gramine-sgx-sign` would probably make it so anyway). This function will need to rely only on the contents of the manifest, not any outside info (esp. untrusted one like the runtime result of `readlink(2)` outside of gramine). 

# options for manifest syntax

## option 1: `type`

```toml
[[sgx.trusted_files]]
uri = "[URI]"
sha256 = "[SHA256]"
type = "symlink"
```

`type = "symlink"` means that `sha256` is over contents of `readlink(2)`, not over the contents of `read(2)`. During runtime, we call `readlink(2)` to untrusted kernel and check if the result matches.

## option 2: `symlink_to`

```toml
[[sgx.trusted_files]]
uri = "[URI]"
symlink_to = "..."
```

The path is symlink. Store the symlink contents directly in the manifest. No `readlink(2)` happens during runtime.

**NOTE:** this option is equivalent to fake bind-mounts, i.e. if you write fake symlink_to, underlying files don't need to exist on the filesystem outside of gramine.

## option 3: allow choice

`uri =` and `type = "symlink"` are mandatory, but whoever is writing manifest can choose if there's `symlink_to = ` or `sha256 = ` (exactly one of those is mandatory, specifying two or none is an error and causes gramine to exit !=0 with message when loading manifest).

### Why Gramine should implement it?

See discussion in gramineproject/gsc#120, but TLDR is, without this feature we're condemned to measure each file under every path it's accessible from. For one, newer distros are usrmerged, which means that there's links like `/bin -> usr/bin`. @dimakuv says on typical distro the size of the manifest explodes 40×.

**Comment by mkow** [comment]
I like that option 1 is more consistent with the rest of our syntax, but I think we should go with option 2, because:
- Less performance overhead (no need to read the host filesystem, less metadata to keep, smaller manifest)
- More readable/debuggable entry in signed manifest - cleartext content instead of an obscure hash. Regular trusted files are hashed only because they are potentially big, so we don't want to embed the contents in the manifest. But symlinks are short, so we can paste them literally.

Although some variance of the option 3 could also be fine, to have that `type = "symlink"` which is nice (but redundant...).

Of course, specifying both `sha256` and `symlink_to` at the same time should be forbidden.

**Comment by dimakuv** [comment]
> @dimakuv says on typical distro the size of the manifest explodes 40×.

Quick remark: this was a misunderstanding. The "40x" factor was for the cases where distros/Docker images have loops due to symlinks (in this case, we collect these paths like `/a/b/c/a/b/...` until Linux hits the limit with `ELOOP`, then GSC stops collecting).

In reality, the increase of some example Docker images can be found in the table here: https://github.com/gramineproject/gsc/pull/120. In other words, anywhere from 2% to 70%.

**Comment by dimakuv** [comment]
How does this proposal help? In particular:

> See discussion in https://github.com/gramineproject/gsc/pull/120, but TLDR is, without this feature we're condemned to measure each file under every path it's accessible from. For one, newer distros are usrmerged, which means that there's links like `/bin -> /usr/bin`.

I currently fail to understand how Gramine will resolve these PAL-level symlinks at runtime. Imagine this scenario:
- We have `sgx.trusted_files = [ {uri = "/usr/bin/bash"}, {uri = "file:/bin", symlink_to = "file:/usr/bin"} ]`
- App inside Gramine calls `open("/bin/bash")`

Now Gramine's PAL needs to resolve `/bin/bash` to the real host path `/usr/bin/bash`. In the manifest's `sgx.trusted_files`, we have all the information needed to resolve and access this file (`/usr/bin/bash`). But what would be the pseudo-code to arrive at this path? PAL will need to do something like this:
1. Iterate over all `sgx.trusted_files` items, find `symlink_to` typed items and compare all prefixes with the prefixes of `/bin/bash`.
2. Iterate again over all `sgx.trusted_files` items, now with the partially-resolved path, repeat step 1 if more nested symlinks.
3. Final iteration (after all possible symlinks are exhausted), to find the regular file among `sgx.trusted_files` items.

This sounds like a very complex logic. Or am I missing a simple solution?

**Comment by mkow** [comment]
No, I'd say that we first go through the trusted files in the manifest and create appropriate metadata structures inside LibOS, similarly to what we do with mounts. This way, when opening a file we go through the path using the metadata (dentries+inodes?) and then to open the final file we query PAL, which will check the hash.
I.e. this symlinks logic should land in LibOS, not PAL. It's a bit ugly though that some of the trusted files will be interpreted by PAL and some by LibOS...

**Comment by dimakuv** [comment]
@mkow Yes, this makes sense. Though it's much more than implied in the issue description.
